[16.0][ADD] stock_release_channel_process_end_time

rousseldenis opened the pull request:
Depends on :

- https://github.com/OCA/wms/pull/493
- https://github.com/OCA/wms/pull/478
- https://github.com/OCA/partner-contact/pull/1464

rousseldenis committed:
[16.0][ADD] stock_release_channel_process_end_date

lmignon commented on stock_release_channel_process_end_date/models/stock_release_channel.py:
```diff
@@ -0,0 +1,35 @@
+
+    _inherit = "stock.release.channel"
+
+    process_end_date = fields.Datetime(
```
@rousseldenis A release channel instance is not created every day. A release channel is a way to organize your pick to deliver process into your company. It's therefore useless to specify the date and time when the process of pickings should be completed. IMO, this process_end_date should be replaced by a process_end_time... 
@jbaudoux What do you thing about?

rousseldenis replied:
> @rousseldenis A release channel instance is not created every day. A release channel is a way to organize your pick to deliver process into your company. It's therefore useless to specify the date and time when the process of pickings should be completed. IMO, this process_end_date should be replaced by a process_end_time...

@lmignon Ok, didn't catch that.

The only drawback I see is that is not flexible (but maybe it's wanted - I don't have the whole flow in mind) as one may want to have channel opened during the day, but pickings done during the night (so, overlap between two days)

lmignon commented on test-requirements.txt:
```diff
@@ -0,0 +1,2 @@
+odoo-addon-stock-release-channel @ git+https://github.com/OCA/wms.git@refs/pull/506/head#subdirectory=setup/stock_release_channel
+odoo-addon-stock-available-promise-release @ git+https://github.com/OCA/wms.git@refs/pull/478/head#subdirectory=setup/stock_available_promise_release
```
```suggestion
odoo-addon-stock-available-to-promise-release @ git+https://github.com/OCA/wms.git@refs/pull/478/head#subdirectory=setup/stock_available_to_promise_release
```

rousseldenis committed:
[IMP] stock_release_channel_process_end_time: Add timezone for end time

As time should be timezoned to correctly compute the real end date, consider
the channel warehouse timezone or company one. If not defined, consider the timezone
as UTC.

rousseldenis commented:
@lmignon This one is ready

lmignon commented on stock_release_channel_process_end_time/models/stock_release_channel.py:
```diff
@@ -0,0 +1,79 @@
+    @api.model
+    def assign_release_channel(self, picking):
+        res = super().assign_release_channel(picking)
+        # Check if a channel has been assigned to the picking and write scheduled_date
```
@rousseldenis For my information, are you sure that a write is triggered if we assign the same value?

rousseldenis replied:
Yes. Checked.

We pass always in that line:

https://github.com/odoo/odoo/blob/16.0/odoo/fields.py#L1320

In fact, I think this is done at purpose to avoid maybe a unneeded read if field value is not in the cache (to be able to compare) and to be able to update the write_uid/write_date as in fact, people try to update the record even if value is the same.

lmignon commented on stock_release_channel_process_end_time/utils.py:
```diff
@@ -0,0 +1,33 @@
+        repl += hours.utcoffset()
+    while repl <= current:
+        repl = repl + timedelta(days=1)
+    return repl
```
@rousseldenis the value must be rounded to the minute. Otherwise the schedule date will not be the same for all the pickings.

rousseldenis replied:
Done

lmignon reviewed (changes requested):
A little change required but work as expected

rousseldenis committed:
[FIX] stock_release_channel_process_end_time: Don't take into account seconds and microseconds

lmignon commented on stock_release_channel_process_end_time/utils.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def next_datetime(current: datetime, hours: time, **kwargs) -> datetime:
+    repl = current.replace(hour=hours.hour, minute=hours.minute, second=0, **kwargs)
```
@rousseldenis and now microsecond...
```suggestion
    repl = current.replace(hour=hours.hour, minute=hours.minute, second=0, microsecond=0 **kwargs)
```

rousseldenis replied:
My god :-)

lmignon reviewed (approved):
LGTM (Code review, functional tests)

lmignon reviewed (changes requested):
~~@rousseldenis The `process_end_date` field is not taken into account into the computation of domains used to retrieve pickings to release | late | done
As consequence, if the process_end_date is computed to tomorrow, the pickings assigned to the channel will not be taken visible since the scheduled_date will be set at tomorrow and the date criteria into the domains is set to today... (see https://github.com/OCA/wms/pull/493/files#diff-a8bb5a8dfb388e49242b1d9fe2274b33dccde4e032d0983647f70fa52399ecbcR174)~~

implemented in https://github.com/OCA/wms/pull/550/commits/8f2038535412a6fc13f34c30793da926d9ee5a73

lmignon commented:
@rousseldenis Changes to the process_end_time is not applied to the process_end_date and pickings to process (maybe it's overkill???)

rousseldenis committed:
[IMP] stock_release_channel_process_end_time: Check if end time is defined before assignation to scheduled_date

lmignon committed:
[FIX] stock_release_channel_process_end_time: releasable picking computation

The compute of releasable picking must take into account the process end date of the stock release channel and not the current datetime.
The way the ORM is working into Odoo prevent to makes query on model with criteria based on values from linked model (comparing column from model A to column of model B). To work arround this limitation, we use a specialized field with a search method. Field's search methods are considered as 'internal' search methods and therefore allows you to use a special 'inselect' operator. This kind operator is usefull since it allows to express a criteria as a plain SQL quey. Thanks to this approach we've an easy way to implement our needs with the Odoo's ORM.

sbejaoui committed:
[FIX] - assign channel end date to picking after release

jbaudoux commented on stock_release_channel_process_end_time/models/stock_picking.py:
```diff
@@ -0,0 +1,72 @@
+        store=False,
+        search="_search_schedule_date_prior_to_channel_process_end_date",
+        help="Technical field to search on not processed pickings where the scheduled"
+        "date is prior to the process end date of available channels.",
```
```suggestion
        help="Technical field to search on not processed pickings where the scheduled "
        "date is prior to the process end date of available channels.",
```

jbaudoux commented on stock_release_channel_process_end_time/models/stock_release_channel.py:
```diff
@@ -0,0 +1,100 @@
+    process_end_time = fields.Float(
+        help="Fill in this to indicates when this channel release process would"
+        "be ended. This information will be used to compute the channel pickings"
+        "scheduled date at channel awaking.",
```
```suggestion
        help="Fill in this to indicates when this channel release process would "
        "be ended. This information will be used to compute the channel pickings "
        "scheduled date at channel awaking.",
```

jbaudoux reviewed (commented):
Some typo in the helps

jbaudoux commented on stock_release_channel_process_end_time/models/stock_release_channel.py:
```diff
@@ -0,0 +1,100 @@
+        company_tz = self.env.company.partner_id.tz
+        for channel in self:
+            channel.process_end_time_tz = (
+                channel.warehouse_id.partner_id.tz or company_tz or False
```
Otherwise float_to_time will fail if tz is a bool
```suggestion
                channel.warehouse_id.partner_id.tz or company_tz or "UTC"
```

sbejaoui committed:
[IMP] - improve process end time location in form view

jbaudoux commented:
@rousseldenis @sbejaoui Can you fix the small remaining issues on this PR ?

lmignon commented:
ping @rousseldenis @sbejaoui

sbejaoui committed:
[FIX] fix help message & add UTC as defauly value for process_end_time_tz

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

sbejaoui commented:
> @rousseldenis @sbejaoui Can you fix the small remaining issues on this PR ?

done

anothingguy committed:
[IMP] add configuration to enable/disable set scheduled date

jbaudoux commented on stock_release_channel_process_end_time/models/res_config_settings.py:
```diff
@@ -0,0 +1,14 @@
+class ResConfigSettings(models.TransientModel):
+    _inherit = 'res.config.settings'
+
+    update_scheduled_date = fields.Boolean(
```
This config is a bit misleading as the picking scheduled date is anyway updated by the release data + delay.

jbaudoux replied:
Could be renamed to stock_release_use_channel_end_date

jbaudoux commented on stock_release_channel_process_end_time/models/res_config_settings.py:
```diff
@@ -0,0 +1,14 @@
+    _inherit = 'res.config.settings'
+
+    update_scheduled_date = fields.Boolean(
+        help="Will update scheduled date of picking based on process end date(time).",
```
```suggestion
        help="Will update scheduled date of picking based on process end date instead of release date + delay.",
```

jbaudoux reviewed (changes requested):
What's the reason behind this new config? In which situation you don't want to plan the work for the given end date?
In my opinion, we should have by default the option active.
If the option is inactive, we should still never have a scheduled date > process end date. So it should still be min(release date + delay, process end date)

jbaudoux commented:
> What's the reason behind this new config? In which situation you don't want to plan the work for the given end date? In my opinion, we should have by default the option active. If the option is inactive, we should still never have a scheduled date > process end date. So it should still be min(release date + delay, process end date)

So the option must remain inactive because it makes the test in `stock_available_to_promise_release` fail

rousseldenis committed:
[IMP] stock_release_channel_process_end_time: Rename configuration parameter

simahawk commented:
@rousseldenis ping (for the build)

simahawk commented:
@rousseldenis do you have time to rebase and shorten/cleanup commits? I can do it otherwise

simahawk commented on stock_release_channel_process_end_time/models/stock_picking.py:
```diff
@@ -0,0 +1,82 @@
+        return [("id", operator_inselect, (query, []))]
+
+    def _after_release_set_expected_date(self):
+        enabled_update_scheduled_date = bool(
```
I'd use str2bool here

simahawk replied:
FTR I tested w/ boolean field tied to a config param: when it sets to False the param is deleted complely -> nothing to fear here.

simahawk reviewed (approved):
LG

rousseldenis commented:
/ocabot merge nobump
